src/condor_starter.V6.1/condor_diagnostic_send_ep_logs.cpp

```diff
@@ -0,0 +1,166 @@
+  const std::filesystem::path & path,
+  std::function<bool(const char *)> filter
+) {
+    std::ifstream i( path );
+    if(! i.good()) {
+        fprintf( stderr, "Failed to open %s\n", path.string().c_str() );
+        return false;
+    }
+
+    std::string line;
+    while( std::getline( i, line ).good() ) {
+        if( filter( line.c_str() ) ) {
+            fprintf( stdout, "%s\n", line.c_str() );
+        }
+    }
+    if(! i.eof()) { return false; }
+
+    return true;
+}
+
+
+// We can't just check for `pattern`, because that will miss lines from the
+// FTO if it's a forked child.  Instead, we look for the first banner line;
+// if it also contains `pattern`, we start including log lines.  We stop
+// including log lines when we find a first banner line that doesn't
+// contain `pattern` (note that we can't check for this _before_ we find
+// our own first banner line, because there are other starters in the log).
+//
+// As this function is currently called, we'll never see the next starter's
+// banner (because this starter hasn't exited yet).
+//
+bool
+relevantPartOfLog( const std::string & pattern, const char * line ) {
+    static bool saw_our_banner_line = false;
+    static bool saw_bad_banner_line = false;
+
+    if( NULL != strstr( line, "******************************************************" )) {
+        bool good_pid = NULL != strstr( line, pattern.c_str() );
+
+        if(! saw_our_banner_line) {
+            saw_our_banner_line = good_pid;
+        } else {
+            saw_bad_banner_line = (! good_pid);
+        }
+    }
+
+    if( saw_our_banner_line && ! saw_bad_banner_line ) {
+        return true;
+    } else {
+        return false;
+    }
+}
+
+bool
+printGlobalStarterLog( pid_t pid ) {
+    //
+    // Acquire the LOG directory.
+    //
+    std::string LOG;
+    if(! param(LOG, "LOG")) {
+        fprintf( stderr, "HTCondor configuration value LOG not defined.\n" );
+        return false;
+    }
+
+
+    //
+    // Acquire NAME from the .machine.ad file.
+    //
+    FILE * file = safe_fopen_no_create( "./.machine.ad", "r" );
+    if( file == NULL ) {
+        fprintf( stderr, "Failed to open .machine.ad.\n" );
+        return false;
+    }
+
+    CondorClassAdFileIterator ccafi;
+    if(! ccafi.begin( file, true, CondorClassAdFileParseHelper::Parse_long )) {
+        fprintf( stderr, "Failed to start machine ad.\n" );
+        return false;
+    }
+
+    ClassAd machineAd;
+    if( ccafi.next( machineAd ) < 0 ) {
+        fprintf( stderr, "Failed to parse machine ad.\n" );
+        return false;
+    }
+
+    std::string NAME;
+    if(! machineAd.LookupString( "Name", NAME )) {
+        fprintf( stderr, "Machine ad did not contain Name attribute.\n" );
+        return false;
+    }
+
+    size_t at = NAME.find_first_of('@');
+    if( at == std::string::npos ) {
+        fprintf( stderr, "Name attribute did not contain @.\n" );
+        return false;
+    }
+    NAME.erase( NAME.find_first_of('@') );
```

Comment:
Not that it really matters, but why not use the `at` variable in `NAME.erase` since we already have it.

Reply:
I'm going to mutter something disgruntled-sounding about "71 commits" instead of actually answering your question, OK?